### ISSUE TITLE
GPU Support

### ISSUE THREAD

**Issue by metawrap-dev** [eduardoleao052/js-torch#1]
It's not an issue - I'm just here to say I love this.

I want to see this get to the point where it will run large transformers. 

For acceleration, you might want to consider something like https://github.com/gpujs/gpu.js, maybe? 

Best of luck.

**Comment by medic-code** [comment]
This is not an area I know much about would love to know more about the approach, @eduardoleao052 is this something you're thinking of using a dependency for or building this from scratch specifically for matrix operations ?

**Comment by eduardoleao052** [comment]
@medic-code after completing the TypeScript support, I'll study the best option to add GPU support. This means that the heavy matrix multiplications will be able to be run on a GPU, speeding up the ML models.

**Comment by metawrap-dev** [comment]
If I may be so bold, I would suggest the following approach. I'm wondering if we can outdo Python if we build from the ground up to focus on GPU integration and performance.  My understanding is Torch and Python mostly have mapped corresponding routines (kernels) in the GPU, so in theory, if we take a from the ground up performance approach, I would like to think we can outperform classical Torch. 

We add decorations inside the Torch-ish objects that allow a visitor object to assemble an entire compute pipeline, kind of a compute abstract syntax tree of what is to be executed that we can map data into.

The pipeline needs to be able to be processed as a whole or chunked along I/O and compute-friendly boundaries to allow us to optimise when running with different GPU availability. 

If the whole pipeline will fit in a big fat GPU then we want it all in there. GPUs are getting bigger and fatter. Let's design for that.  If we need to break it up into discrete steps with a small or multiple small GPUs, then we need to be able to do that as well. 

```                                                                                                                                                                                                                       
 ┌────────────┐    ┌────────────┐     ┌──────────┐     ┌──────────────┐     ┌─────────────┐    
 │  Torchish  │    │ Pipline    │     │ Pipleine │     │ Orchestrator │     │ GPU         │    
 │   Objects  ├──► │ Assembler  ├───► │   AST    ├────►│              ├───► │ Gateway     │    
 └────────────┘    └────────────┘     └──────────┘     └──────────────┘     └─────────────┘    
                                                                                               
                                                                                               
  The executing      Can take        Representation      Takes the            Takes chunks     
    object           the planned     of what needs       pipeline and         of the AST       
                     execution and   to be computed      determines how       and data and     
                     turn it into                        it will be executed  executes in GPU  
                     Pipeline AST                        against the GPU                       
                                                                                               
                                                                                                                                                                                                                                                                                                                                                                          
```

Make use of ReadableStream<X> WritableStream<X> TransformerStream<X> for compatibility with browser and node and everything async. We need to stream everything to minimise memory usage. We need to be able to stream data and models from storage into the GPU and then data out without maxing out v8's limited memory. 

As a kind of proof of concept and possibly even the solution, for version 0 we take the AST for the pipeline and build a function that can be passed into https://github.com/gpujs/gpu.js Let it do the heavy lifting for now. If we need to do something else in the future, we can replace this step with something else and keep the rest of the design and implementation.

**Comment by eduardoleao052** [comment]
@metawrap-dev I like the spirit, and you clearly have a great understanding of hardware manipulation on Pipelines. I think we could make it work, and adding at least the `gpu.js` is a clear first step in my opinion. If a simple GPU acceleration is put in place, it could already allow a much wider range of applications for the library.

**Comment by metawrap-dev** [comment]
Thanks!

To give a toy example: `add(mult(w,x), mult(y,z))`

We really want to build an object tree that represents the promise to execute this. 

That way, instead of passing  `(w * x)`  and then `(y  * z)` and then waiting for the result of these and then passing `r1 + r2` to the GPU, we could more efficiently pass `(w * x) + (y * z)` in one go and get one value back. 

So we go from 

```mermaid
sequenceDiagram
    participant JavaScript
    participant GPU
    JavaScript->>GPU: w * x
    GPU->>JavaScript: r1
    JavaScript->>GPU: y * z
    GPU->>JavaScript: r2
    JavaScript->>GPU: r1 + r2
    GPU->>JavaScript: a
```

to

```mermaid
sequenceDiagram
    participant JavaScript
    participant GPU
    JavaScript->>GPU: (w * x) + (y * z)
    GPU->>JavaScript: a
```

This will work efficiently when we pass a torch-ish object into another's method as a parameter 

Not much if we pass in just a scalar/primitive unless we make a special class, eg a Number class that would allow a whole tree to be linked. We would, of course, make it as granular as possible (no point in building the internal AST of a matrix operation). 

If the library is skewed towards being able to build the largest connected Pipeline tree, then we have the ability to pass the whole intention into the GPU and optimizers and have them work their magic.

**Comment by eduardoleao052** [comment]
Cool! I could see that working building upon the `Tensor` and `Module` objects. The `Module.forward()` method is always spelled-out before the model is actually run. Then we could build this "graph" with all the operations, and send them in parallel to the GPUs!

I'm currently finishing to add TypeScript support to the library, and will push it to main soon (currently it's in the `develop` branch).

Would you by any chance be willing to work on a PR, so that we can try to implement the `GPU.js` support? I'd love to take on that challenge!

**Comment by metawrap-dev** [comment]
Sure! Once you complete the TypeScript support. I do a lot of prototyping in TypeScript, so it's my happy place. 

In the meantime, I will look into implementing the toy example above and try out some design ideas.

**Comment by metawrap-dev** [comment]
These two features in `gpu.js` are likely going to be load-bearing for us. 

https://github.com/gpujs/gpu.js/blob/develop/README.md#combining-kernels
https://github.com/gpujs/gpu.js/blob/develop/README.md#create-kernel-map

Some notes here for things I want to investigate. 

1. How to combine async/Promise and direct reference. (P,R) or (R{P})
1. Data vs Function - how to model so we maximise "streamability"
1. How do we stream "big" data in the pipeline? Chunk data along with execution nodes? Transform pipeline like map/reduce repeating F with D1...DN then combining (where)?
1. How to chunk the pipeline graph? Assuming not chunking into big GPU echoey is the easy case. Pinch the pipe at the "thinnest" place. Define "thinnest" with the dimensions we have.
1. Future - Is it possible to extend gpu.js to support "Persistent Kernels"
1. Object reference equivalence optimisation?
1. Object hash equivalence optimisation? (overkill)
1. It feels like being monad-like is going to play a big part of this working, so make everything "complete".

**Comment by eduardoleao052** [comment]
That's awesome, looks like a great start!
Btw I'll merge the TypeScript branch into main today, I'll let you know.

**Comment by eduardoleao052** [comment]
@metawrap-dev done! Merged the TS into the main branch

**Comment by metawrap-dev** [comment]
Thanks. I will take a look. If you have any typescript questions I'm happy to help there as well. 

I'm going to continue in a different repository until I have the major questioned I outlined above answered. 

Today I'm going to tackle the streaming and batching big data question. Big as in more than can fit into memory. The goal (I) will be to have a toy versions of a model load with chained to execution of operations and then output. 

I want to be able to get to the point where I execution is triggered only when we want to save the result. So, complete lazy loading and execution with the ability to inspect the pipeline graph. 

Next after that the goal (ii) will be to have a combination of granularities so we have some operations that will be handled by a prepositioned kernel and then some that are weaved together with permutations in code. 

At that point it will be possible to create a visitor class that can analyse and chop/chunk and transpile the pipeline graph.

I can probably only get some of (i) done today given the time I have as just coding up the testing and plumbing is going to be a majority of the work.

**Comment by eduardoleao052** [comment]
Cool! If you want any help at some point, just let me know. About the TypeScript, the only issue left to solve are a couple of `type: any`s and unused variables. Just a matter of keeping the code cleaner.

**Comment by metawrap-dev** [comment]
If you can avoid `any`, it would be a Good Thing™️. 

If it is the case where you don't know the type yet in the flow of execution and it will be resolved later on, then using `unknown` rather than `any` is preferred. 

If it is just a case of trying to get things to match and `any` is used to get past that, then I can definitely help with that. 

Here is my work so far. 

https://github.com/metawrap-dev/streaming-compute-poc

**Comment by metawrap-dev** [comment]
The DataNumber class is just a start and just a minimalist toy model. Yes, higher order objects like tensors will be modelled the same way. I'm just starting very small as there is a lot of type semantics to work out to make everything composable before I move onto something more complex.

**Comment by metawrap-dev** [comment]
I've updated the repository. I've also implemented some more complex composition involving sources. However, I'm still only using primitive number types and the multiply operation for now. My goal is to get the streaming working before I start doing anything more complex or more granular. 

![image](https://github.com/eduardoleao052/js-torch/assets/42017256/b584ae33-9ac3-4b53-879d-cf3161b22d89)

**Comment by metawrap-dev** [comment]
In-memory `Source` and `Destination` added.  I didn't to tackle I/O and streaming yet but this models a lot of the problem and solution in memory already. 

`{DestinationMemory(0 stored, 1 in buffer, 5 batch size) <= [( multiply {SourceMemory(4 elements, 0 index, 1 batch size) <= [10,10,10,10]} => unresolved )]=>[]}'`

**Comment by metawrap-dev** [comment]
Gathering some more thoughts for (i)

I'm pretty close to being able to serialise and deserialise the execution state. 'Source' and 'Destination' batch themselves now. What is missing is generalising this to the 'Compute' elements. I'm going to tease out the internal run-state of each element into something common that can be marshalled.

What will that get us? 

1. Batching and chunking of execution where possible. 
1. Execution back-pressure signalling on the running pipeline can self optimise. 
1. Able to freeze and store long running execution to storage and later on thaw and continue execution from where it left off.

I'm also thinking about adding an unrolling strategy as the same kind of parameter as batch size. With this, the system that is trying to optimise the pipeline graph will also be able to experiment with unrolling.

And that implies that metrics are going to needed. 

The optimising system observing the execution needs to be able get metrics at each point in the pipeline and maintain some previous data points to make a comparison. Enough to give it the ability to converge but not too much that it overwhelms the task at hand. 

1. Common 'Strategy' object? Something that can be introspected and modified even just using dumb random search. Element can take a strategy and apply it to State. 
1. Common 'State' object? Something that reflects the concrete application of 'Strategy' or separate 'Config' for this? State = Runtime + Config?

**Comment by metawrap-dev** [comment]
Implemented `Strategy`, `Config`, and `State` across all the elements as described above. `Strategy` needs to be put into practice, so what is there is a placeholder for now. 

![image](https://github.com/eduardoleao052/js-torch/assets/42017256/1b437c15-4f4a-4530-8fad-ea8b0d1c6535)

So close to 100% :)

Tentative flow control so that operations asynchronously [will block until data becomes available](https://github.com/metawrap-dev/streaming-compute-poc/blob/fa143715885df896320bee9e14e0ebbde0337354/__tests__/DestinationMemory.test.ts#L71). 

It is time to add more complex `Data` and `Compute` elements to refine the design as single number flow seems done. I've exhausted what I can do with these now. 

The goal is to make sure that execution only occurs [after resolve](https://github.com/metawrap-dev/streaming-compute-poc/blob/fa143715885df896320bee9e14e0ebbde0337354/__tests__/DestinationMemory.test.ts#L199), but there are situations were populating `Source` or `Destination` will [start to trigger flows](https://github.com/metawrap-dev/streaming-compute-poc/blob/fa143715885df896320bee9e14e0ebbde0337354/__tests__/DestinationMemory.test.ts#L37). 

The design looks like it is holding together and will do what I want. More complex objects should now shake any bugs or design issues with the way I want data to flow and obey batch/chunk sizes. It should will allow me to test the roll/unroll strategy ideas, as I need a Compute element with at least a few layers of execution to support that.

**Comment by metawrap-dev** [comment]
Thanks,. 

Yes, now that I have the flow and semantics thought out and lazy execution working, more complex objects are now needed. 

Just thinking some things out loud again here.

Code-Gen/Optimizer: 

1. Just take the pipeline graph (PG), break it into sub-graphs of the right size and transform each sub-graph into a GPU kernel. 
2. "Where" it is broken up will be affected by `Strategy`. At a point that maximises throughput. Treat the GPU as I/O.
3. "How" we size..  Good question.
4. If the same PG "shape" is being built, executed and then torn down, over and over, then how do we associate with past metrics? 
5. Need a hash that bubbles up to the root that reflects the underlying execution pattern. Each identical sub-graph will have same hash. Cost worth it? Will see? Likely yes. 
6. Add metrics now. Too important to leave until later. 
7. How do we cleanly associate optimiser with compute element without using globals. 
8. Some 'Compute' elements will have 1:1 correspondence with inbuilt (hardware ops) and pre-positioned (software) kernels. Noting inside. Just an Element. 
9. Some 'Compute` elements will have control over unrolling based on 'Strategy`. Build PG will depend on `Strategy1
10. Can we model flow-based/Persistent computation in the design now? Build a pipeline that you feed data into a source and it feeds into a `Destination`? Reset vs Revisit resolve()/Resolved semantics.  We may need to mock behaviour for now, but it is important that this is designed in at ground level now.
11. Naive vs Easy number packing? Is it better to store the matrix in 1d or 2d array if the whole flow from coder to GPU is taken into account? Not obvious. Measure. 

Mock `Compute` and `Data` elements.

1. DataVectorN
1. DataMatrixMxN
1. ComputeDotN (Dot Product)
1. ComputeDot4GPU (Dot Product GPU)
1. ComputeAverage (Average Pooling)
1. ComputeAverage4x4GPU (Average Pooling) (Hardware Op)
1. ComputeMultiplyMatrix
1. ComputeMultiplyMatrix4x4GPU

This should give enough. The plan for today is just to get the new mock data elements in. Probably all I have time for. 

I suspect the design won't survive the first data element without changes, but here goes. : )

**Comment by metawrap-dev** [comment]
`DotProduct4`

A lot to unpack here. 

![image](https://github.com/eduardoleao052/js-torch/assets/42017256/beaac0b1-2f91-4159-89f1-a724a475ffdb)

Great success.. but.. 

From a flow aspect, there are several ways to computer dotN. Not all of them are obviously nonsense. Also this is going to be the same for all operations. 

1. Supply two vectors. Output a single number. OK.
2. Supply an array of two vectors, batch compute and return a vector (Ok.. makes sense.. we want that)
3. Stream in two huge vectors of known same length. Stream compute. Output a single number. (Nice) 
4. Same as 3 but we arbitrarily decide when to cut the sausage and output a number. (Can't really think of a use case)
5. Hold my beer. Stream in a stream of huge vectors. Stream compute. Stream out vector of numbers. (Very nice)
6. Same as 5 but we arbitrarily decide when to cut the sausage and output a number. (Can't really think of a use case, or how unless we single end of vector somehow? ).

This adds a lot of fragile complexity in having generalized arguments. 

I'm wondering if I can create IStreamify<T> so to formally transform non-streaming to a streaming version. Otherwise, It's obvious I am going to end up in array hell. (Is it a matrix or an array of vectors to streaming output of vector etc. )

Time to dive deep into some of the latest typescript features.

**Comment by metawrap-dev** [comment]
TypeScript's type system is doing some heavy lifting here. 

```TypeScript
type v4 = Vector<number, 4> | DataVectorN | (number | IData<number>)[]

const source = new SourceMemory([1, 1, 1, 1], new DataVectorN([1, 1, 1, 1]), [1, new DataNumber(1), 1, 1], [1, 1, 1, 1])

const compute = new ComputeLength4()

const streamer = new Streamer<v4, number>(source, compute)
```

**Comment by metawrap-dev** [comment]
Some more thoughts. 

Streaming in such a way that engages elegantly with TypeScript's type system but remains performant; There are few ways to go. 

The main issue is argument cardinality and the marshalling stage into an `Compute` element with different argument sizes. 

Either the `Streamer` needs to know how many arguments it is dealing with, or the `Compute` element needs to know.

Source(2)=>Streamer(2)=>Compute(2)

`ICompute<I,N,O>`

This needs to be expressed in the type, the type system will flag a mismatch so N needs to propagate properly. It will be used to drive Vector<N> but we also want it as a cardinal for use at runtime. 

The code is bit of a mess right now because I have a few ideas in play simultaneously and not all of them will be the winner. 

Streamer vs no Streamer: `Compute` input becomes `Vector` vs Source.

Sausage Machine vs Structured. 

Sausage machine (or some variant) may actually be the best way forward and the most flexible. 

This is all about being able to chunk up code and data into big blobs that we can blat into the GPU without a lot of back and forth. Ultimately all the design decisions are driven by that goal.

**Comment by medic-code** [comment]
@metawrap-dev its really fascinating to see your thought process, I've just been wondering if you had any ideas if there's a better way to document this other than an open issue ? 

For example maybe an open PR with a readme.md, or a wiki page or something ? It's okay if you feel this way of documenting your thoughts is fine. I just imagine that after a while you might want to collect your thoughts in a summary and rationale so far. I realise that this is still a part of discovery as you go and you may want to postpone documenting in a more formalised way. Just some thoughts.

**Comment by metawrap-dev** [comment]
Yep, eventually. At the moment, I'm still experimenting. I assumed here was good because others could chip in. It just kind of happened organically as I write down most of these thoughts while mobile and then refer back to them as I work, so very convenient for me. 

I will gather my thoughts elsewhere.

**Comment by medic-code** [comment]
I'm definitely not suggesting you should gather your thoughts else where outside of github, as i said its very useful to see your thought process/allow others to contribute. I would imagine that you may find trying to scroll back through more than 29 posts a little more challenging as you keep documenting, that is all.

**Comment by metawrap-dev** [comment]
29 (now 30) posts is fine. Scrolling is fine. That's not a long issue, [this](https://jira.atlassian.com/browse/JRACLOUD-63150) is a long issue. 😁 But we finally got there. 

Hopefully, I will be finished before this gets that long. 

I have the types mostly tamed. 

For the initial toy system. I have a generalised way of defining a compute element. The pipeline can be constructed out of resolved (static) or unresolved (`Compute`, `Source` or data with unresolved elements. )

I have temporarily removed `Streamer` as I think I need something more general.

**Comment by metawrap-dev** [comment]
I have a naive type signature implementation that will compose. In this form the generator parameters can't be completely inferred from the input types as I am trying to be flexible on the output stage.  I will probably need to remove this flexibility and always return a standard form that can be derived from the source type.

```TypeScript
 it('Generator matching 1:1 compute signature', async () => {
    // A source of pairs matching signature of compute element.
    const source = new SourceMemory<number, 4, 2>(
      [
        [10, 10, 10, 10],
        [10, 10, 10, 10],
      ],
      [
        [10, 10, 10, 10],
        [10, 10, 10, 10],
      ],
    )

    // The compute element that will take input from the source
    const compute = new ComputeAddV4()

    // The generator
    const generator = new Generator<number, 4, 2, number, 4, 2, number, 4, 1, number, 4, 1>(source, compute)

    // If the generator parameters can be inferred from inputs then we can have this
    // const generator = new Generator(source, compute)

    // input pairs       output vectors
    // <number,4,2>[] => <number,4,2>[] = Output<number,4,2>,1,0

    // This generator effectively becomes a new source or can we make it appear to be a new compute element?
    // Does it matter as long as it has a type and matches signature of IResolves<Output<number,4,2>,1,0>
    // Which is the type integral of the source's output.

    expect(generator).toBeDefined()
  })
```

**Comment by metawrap-dev** [comment]
`Generator` looks like it will work. In the current experiment, I am looking at still allowing `Source` to work for individual arguments and `Generator` for collated sets of arguments. I am tightening up the types with a view to simplifying the classes. 

The required code in a `Compute` element has been slashed. A bit more type and base class algebra should get it down to something much more manageable than it was when I started. 

Generators that need to transform the `Source` to suit the `Compute` element as well as `Compute` elements with heterogeneous arguments will be the next challenge.  Varadic Tuple arguments and partial argument consumption may work for the latter. I have an idea. Still thinking about how to finesse this with the generator such that the type system performs the maximum amount of lift. What will come out of this, at least, is a proposal with a concrete example for TypeScript to support materialising value instance type generics arguments accessible as constants. 

I am backtracking on horizontal vectors as they add a disruptive fog to singleton vs vector vs spread argument binding as I am going through the process of harmonising all the code and moving it into base classes.  I will add them back if I can make everything else work or will sacrifice them if they are too hard to keep.  I would like them to contribute to optimization knowledge as one would typically get better CPU/GPU cache performance over a contiguous, non-striding source. I may try to hold that knowledge in the object itself in the `Strategy` if I can't keep it at a type/class level.  

In the meantime, plan B is `Compute2` `Compute3` `Compute4`, etc, where N is the number of types supported by the compute element.

**Comment by metawrap-dev** [comment]
Toy pipeline with `Generator` seems to work, Just with matching inputs for now. A `Generator` element will take a source, pass it through a single `Compute` element and generate a new `Source` that contains all the output of that process. 

If I create a `Generator`

```
{Generator(0 elements, 0 atoms, 0 index, 1 batch size){
  SourceMemory(3 elements, atoms 3, 0 index, 1 batch size) 
    <= [[[1,1,1,1],[2,2,2,2]],[[3,3,3,3],[5,5,5,5]],[[7,7,7,7],[11,11,11,11]]]}
        =>ComputeAddV4=>[]}
```

.. and then write it to a `DataStore` it sucks the data out from the `Generator (as a `Source`), triggering `resolve()` down the chain and we get..  

```
{DestinationMemory(3 stored, 0 in buffer, 1 batch size) 
   <= []
      =>[[3,3,3,3],[8,8,8,8],[18,18,18,18]]}
```

So, I'm going to keep things simple now and try to add a stepping stone towards code generation. 

This will be like a `Generator` but will take a sequence of `Compute` elements in a complex array. In effect a multidimensional pipeline. It will in itself be a compute element. Or I may be limited to just composing two at a time. I will see how far I can push the types system,.

**Comment by metawrap-dev** [comment]
Super busy at the moment. Hopefully back into this soon.

**Comment by metawrap-dev** [comment]
Also relevant. 

https://github.com/HigherOrderCO/Bend

I was leading towards a similar tree-based flow.  I'm going to study this and see what can be gleaned.

**Comment by metawrap-dev** [comment]
Here I was thinking that the way I was approaching some of the composition was weird and bordering unacceptable, but it appears that WebGPU does something similar. I will borrow some terms from here.  :D

https://google.github.io/tour-of-wgsl/types/vectors/constructors/
